src/agent/src/image_rpc.rs

```diff
@@ -0,0 +1,160 @@
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use protocols::image;
+use tokio::sync::Mutex;
+use ttrpc::{self, error::get_rpc_status as ttrpc_error};
+
+use crate::sandbox::Sandbox;
+//use crate::trace_rpc_call;
+use crate::rpc::CONTAINER_BASE;
+
+const SKOPEO_PATH: &str = "/usr/bin/skopeo";
+const UMOCI_PATH: &str = "/usr/local/bin/umoci";
+
+// Convenience macro to obtain the scope logger
+macro_rules! sl {
+    () => {
+        slog_scope::logger()
+    };
+}
+
+pub struct ImageService {
+    sandbox: Arc<Mutex<Sandbox>>,
+}
+
+impl ImageService {
+    pub fn new(sandbox: Arc<Mutex<Sandbox>>) -> Self {
+        Self { sandbox }
+    }
+
+    fn pull_image_from_registry(image: &str, cid: &str, source_creds: &Option<&str>) -> Result<()> {
+        let source_image = format!("{}{}", "docker://", image);
+
+        let manifest_path = format!("/tmp/{}/image_manifest", cid);
+        let target_path_manifest = format!("dir://{}", manifest_path);
+
+        // Define the target transport and path for the OCI image, without signature
+        let oci_path = format!("/tmp/{}/image_oci:latest", cid);
+        let target_path_oci = format!("oci://{}", oci_path);
+
+        fs::create_dir_all(&manifest_path)?;
+        fs::create_dir_all(&oci_path)?;
+
+        info!(sl!(), "Attempting to pull image {}...", &source_image);
+
+        let mut pull_command = Command::new(SKOPEO_PATH);
+        pull_command
+            .arg("--insecure-policy")
+            .arg("copy")
+            .arg(source_image)
+            .arg(&target_path_manifest);
+
+        if let Some(source_creds) = source_creds {
+            pull_command.arg("--src-creds").arg(source_creds);
+        }
+
+        let status: ExitStatus = pull_command.status()?;
+
+        if !status.success() {
+            return Err(anyhow!(format!("failed to pull image: {:?}", status)));
+        }
+
+        // Copy image from one local file-system to another
+        // Resulting image is still stored in manifest format, but no longer includes the signature
+        // The image with a signature can then be unpacked into a bundle
+        let status: ExitStatus = Command::new(SKOPEO_PATH)
+            .arg("--insecure-policy")
+            .arg("copy")
+            .arg(&target_path_manifest)
+            .arg(&target_path_oci)
+            .arg("--remove-signatures")
+            .status()?;
+
+        if !status.success() {
+            return Err(anyhow!(format!("failed to copy image: {:?}", status)));
+        }
+
+        // To save space delete the manifest.
+        // TODO LATER - when verify image is added, this will need moving the end of that, if required
+        fs::remove_dir_all(&manifest_path)?;
+        Ok(())
+    }
+
+    fn unpack_image(cid: &str) -> Result<()> {
+        let source_path_oci = format!("/tmp/{}/image_oci:latest", cid);
+        let target_path_bundle = format!("{}{}{}", CONTAINER_BASE, "/", cid);
```

Comment:
We should really be using `Path` to build up paths.

Reply:
It's easier to use String for the arguments you want afterwards. As far as I know, Path conversion to str may introduce a few more errors to handle.

Reply:
I appreciate it's easier, but consider what would happen if `cid` was "`../../../../sbin/init`" or something crazy like that? In fact [that isn't allowed](https://github.com/kata-containers/kata-containers/blob/main/src/agent/src/rpc.rs#L113) but you get the idea ;)

Reply:
Your are right, `verify_cid` is added at the begin of PullImage.

Reply:
@wllenyj I agree with @jodh-intel here. Could we use `Path` to more safely build those paths and convert them to strings when needed? It's slightly less easy (you'll need an additional `ok_or_else()`) but I think it's worth it.

Reply:
fixed